rs/registry/node_provider_rewards/src/v1_rewards.rs

```diff
@@ -0,0 +1,535 @@
+    days_in_period: u64,
+    rewards_table: &NodeRewardsTable,
+    subnet_metrics: HashMap<PrincipalId, Vec<NodeMetricsHistoryResponse>>,
+    rewardable_nodes: &[RewardableNode],
+) -> RewardsPerNodeProvider {
+    let mut rewards_per_node_provider = HashMap::default();
+    let mut rewards_log_per_node_provider = HashMap::default();
+
+    let mut all_assigned_metrics = daily_node_metrics(subnet_metrics);
+    let subnets_systematic_fr = systematic_fr_per_subnet(&all_assigned_metrics);
+    let node_provider_rewardables = rewardables_by_node_provider(rewardable_nodes);
+
+    for (node_provider_id, node_provider_rewardables) in node_provider_rewardables {
+        let mut logger = RewardsLog::default();
+        logger.add_entry(
+            LogLevel::High,
+            LogEntry::CalculateRewardsForNodeProvider(node_provider_id),
+        );
+
+        let assigned_metrics: HashMap<PrincipalId, Vec<DailyNodeMetrics>> =
+            node_provider_rewardables
+                .iter()
+                .filter_map(|node| {
+                    all_assigned_metrics
+                        .remove(&node.node_id)
+                        .map(|daily_metrics| (node.node_id, daily_metrics))
+                })
+                .collect::<HashMap<PrincipalId, Vec<DailyNodeMetrics>>>();
+        let node_daily_fr =
+            nodes_idiosyncratic_fr(&mut logger, &assigned_metrics, &subnets_systematic_fr);
+
+        let rewards = node_provider_rewards(
+            &mut logger,
+            &node_provider_rewardables,
+            node_daily_fr,
+            days_in_period,
+            rewards_table,
+        );
+
+        rewards_log_per_node_provider.insert(node_provider_id, logger);
+        rewards_per_node_provider.insert(node_provider_id, rewards);
+    }
+
+    RewardsPerNodeProvider {
+        rewards_per_node_provider,
+        rewards_log_per_node_provider,
+    }
+}
+
+/// Computes the idiosyncratic daily failure rates for each node.
+///
+/// This function calculates the idiosyncratic failure rates by subtracting the systematic
+/// failure rate of the subnet from the node's failure rate for each day.
+/// If the node's failure rate is less than the systematic failure rate, the idiosyncratic
+/// failure rate is set to zero.
+fn nodes_idiosyncratic_fr(
+    logger: &mut RewardsLog,
+    assigned_metrics: &HashMap<PrincipalId, Vec<DailyNodeMetrics>>,
+    subnets_systematic_fr: &HashMap<(PrincipalId, TimestampNanos), Decimal>,
+) -> HashMap<PrincipalId, Vec<Decimal>> {
+    let mut nodes_idiosyncratic_fr: HashMap<PrincipalId, Vec<Decimal>> = HashMap::new();
+
+    for (node_id, daily_metrics) in assigned_metrics {
+        let failure_rates = nodes_idiosyncratic_fr.entry(*node_id).or_default();
+
+        for metrics in daily_metrics {
+            let systematic_fr = subnets_systematic_fr
+                .get(&(metrics.subnet_assigned, metrics.ts))
+                .expect("Systematic failure rate not found");
```

Comment:
If we're running this in registry, and we panic, we would not be able to calculate any NP rewards.  I'm wondering if this is what's desired here.  If that's the best response, then this might be what we want...